src/block/inventory/EnchantInventory.php

```diff
@@ -23,18 +23,101 @@
 
 namespace pocketmine\block\inventory;
 
+use pocketmine\data\bedrock\EnchantmentIdMap;
 use pocketmine\inventory\SimpleInventory;
 use pocketmine\inventory\TemporaryInventory;
+use pocketmine\item\enchantment\EnchantmentHelper as Helper;
+use pocketmine\item\enchantment\EnchantmentInstance;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\item\VanillaItems;
+use pocketmine\network\mcpe\protocol\PlayerEnchantOptionsPacket;
+use pocketmine\network\mcpe\protocol\types\Enchant;
+use pocketmine\network\mcpe\protocol\types\EnchantOption;
 use pocketmine\world\Position;
+use function array_map;
+use function array_merge;
+use function count;
+use function is_null;
 
 class EnchantInventory extends SimpleInventory implements BlockInventory, TemporaryInventory{
 	use BlockInventoryTrait;
 
 	public const SLOT_INPUT = 0;
 	public const SLOT_LAPIS = 1;
 
+	/** @var EnchantOption[] $options */
```

Comment:
Okay, but we need to create Enchant, EnchantOption somewhere to send the packet. If it is created only in EnchantmentHelper, will it be ok? Then I'll make EnchantmentHelper::getEnchantOptions return an array of EnchantmentInstance arrays instead of EnchantOption, and create EnchantmentHelper::sendEnchantOptions where the packet will be sent to the player

Reply:
The protocol types should not be stored. They should be created as-needed to fill the appropriate packets.

Reply:
So what I'm saying is that I'll make sure I don't store Enchant and EnchantOption instances anywhere. Instead, EnchantmentInstance (which is not a protocol type) will be stored. But Enchant and EnchantOption will be created in EnchantmentHelper to send the player's EnchantOptionsPacket. Is that fine?